packages/@aws-cdk/aws-redshift/lib/privileges.ts

```diff
@@ -0,0 +1,133 @@
+import * as cdk from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { DatabaseProps } from './database-props';
+import { DatabaseQuery } from './private/database-query';
+import { TablePrivilege as SerializedTablePrivilege, UserTablePrivilegesHandlerProps } from './private/handler-props';
+import { ITable } from './table';
+import { IUser } from './user';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct as CoreConstruct } from '@aws-cdk/core';
+
+/**
+ * The Redshift table and action that make up a privilege that can be granted to a Redshift user.
+ */
+export interface TablePrivilege {
+  /**
+   * The table on which priveleges will be granted.
+   */
+  readonly table: ITable;
+
+  /**
+   * The actions that will be granted.
+   */
+  readonly actions: TableAction[];
+}
+
+/**
+ * An action that a Redshift user can be granted privilege to perform on a table.
+ */
+export enum TableAction {
+  /**
+   * Grants privilege to select data from a table or view using a SELECT statement.
+   */
+  SELECT,
+
+  /**
+   * Grants privilege to load data into a table using an INSERT statement or a COPY statement.
+   */
+  INSERT,
+
+  /**
+   * Grants privilege to update a table column using an UPDATE statement.
+   */
+  UPDATE,
+
+  /**
+   * Grants privilege to delete a data row from a table.
+   */
+  DELETE,
+
+  /**
+   * Grants privilege to drop a table.
+   */
+  DROP,
+
+  /**
+   * Grants privilege to create a foreign key constraint.
+   *
+   * You need to grant this privilege on both the referenced table and the referencing table; otherwise, the user can't create the constraint.
+   */
+  REFERENCES,
+
+  /**
+   * Grants all available privileges at once to the specified user or user group.
+   */
+  ALL
+}
+
+/**
+ * Properties for specifying privileges granted to a Redshift user on Redshift tables.
+ */
+export interface UserTablePrivilegesProps extends DatabaseProps {
+  /**
+   * The user to which privileges will be granted.
+   */
+  readonly user: IUser;
+
+  /**
+   * The privileges to be granted.
+   *
+   * @default [] - use `addPrivileges` to grant privileges after construction
+   */
+  readonly privileges?: TablePrivilege[];
+}
+
+/**
+ * Privileges granted to a Redshift user on Redshift tables.
+ */
+export class UserTablePrivileges extends CoreConstruct {
```

Comment:
From what I understand here, it seems that this construct is managing the grants for a single user to multiple tables in a single construct (and hence, custom resource). Yet, the user is free to create another construct of this type for the same user and manage another set of grants.

What happens if these two sets of grants for the same user collide?

Would it be simpler instead to manage user-table mapping as a single construct? For every user-table combination, a new construct will be created.

Reply:
I made UserTablePrivileges module-private to help avoid collision via manual creation of the construct.

Since only one UserTablePrivileges construct is created per UserBase, the risk here is if a user is referenced in two different locations (one created/import, the other imported) and granted privileges to the same action on the same table. Then, if one of those grants is revoked, the custom resource will remove the privilege from the user even though the other reference to the user still thinks that it has privileges to the table on that action. I don't see any reliable way of managing this risk, is it acceptable?

Reply:
I guess we could refuse to grant privileges if we detect that they already exist?

Reply:
Discussed offline. Let's go ahead with the current impl and see how this pans out.